libusb/os/windows_winusb.c

```diff
@@ -2017,6 +2017,186 @@ static int winusb_get_device_list(struct libusb_context *ctx, struct discovered_
+	}
+
+	for (size_t k = 0; k < src_length; k += 2) {
+		codepoint = *src++;
+		codepoint |= ((uint16_t)*src++) << 8;
+		if (codepoint <= 0x7f) {
+			if (count < dst_length) {
+				dst[count] = (char)codepoint;
+			} else {
+				overflow = true;
+			}
+			count += 1;
+			if (0 == codepoint) {
+				--count;  // add null-terminator below
+				break;
+			}
+		}
+		else if (codepoint < 0x7ff) {
+			if ((count + 1) < dst_length) {
+				dst[count] = (char)(0xC0 | (codepoint >> 6));
+				dst[count + 1] = (char)(0x80 | (codepoint & 0x3F));
+			} else {
+				overflow = true;
+			}
+			count += 2;
+		}
+		else {
+			if ((count + 2) < dst_length) {
+				dst[count] = (char)(0xE0 | (codepoint >> 12));
+				dst[count + 1] = (char)(0x80 | ((codepoint >> 6) & 0x3F));
+				dst[count + 2] = (char)(0x80 | (codepoint & 0x3F));
+			} else {
+				overflow = true;
+			}
+			count += 3;
+		}
+	}
+
+	if (count < dst_length) {
+		dst[count] = 0;
+	}
+	++count;  // include null terminator
+
+	if (overflow && dst_length) {
+		// truncate respecting UTF-8 character boundaries
+		size_t idx = dst_length - 1;
+		while (idx && (0x80 == (dst[idx] & 0xC0))) {  // utf-8 continuation byte
+			--idx;
+		}
+		dst[idx] = 0;
+	}
+
+	return (int) count;
+}
+
+/*
+ * Backend implementation for libusb_get_device_string().
+ * 
+ * Windows makes getting the common device strings
+ * very difficult.  DEVPKEY_Device_* does not have SerialNumber,
+ * and it reports the driver manufacturer, not the device manufacturer.
+ * 
+ * We could parse the serial number from the DEVICE_ID string:
+ * https://learn.microsoft.com/en-us/windows-hardware/drivers/install/device-instance-ids
+ * https://learn.microsoft.com/en-us/windows-hardware/drivers/install/standard-usb-identifiers
+ * 
+ * However, using the dev_id for getting the serial number is 
+ * definitely not recommended.
+ *
+ * The following implementation uses an IOCTL
+ * to the parent USB hub to perform the USB control request for the
+ * string descriptor without opening the USB device.
+ * While we would rather not invoke USB IO, we currently lack a 
+ * better option.
+ */
+static int winusb_get_device_string(libusb_device* dev,
+	enum libusb_device_string_type string_type, char* data, int length)
+{
+	struct winusb_device_priv* priv = usbi_get_device_priv(dev);
+	struct libusb_context* ctx = DEVICE_CTX(dev);
+	DWORD size;
+	DWORD ret_size;
+	struct string_descriptor_req_s sd;
+
+	if ((NULL != data) && (length > 0)) {
+		*data = 0;
+	}
+	if (NULL == dev->parent_dev) {
+		return LIBUSB_ERROR_NOT_SUPPORTED;
+	}
+
+	uint8_t string_descriptor_idx;
+	switch (string_type) {
+	case LIBUSB_DEVICE_STRING_MANUFACTURER: string_descriptor_idx = dev->device_descriptor.iManufacturer; break;
+	case LIBUSB_DEVICE_STRING_PRODUCT: string_descriptor_idx = dev->device_descriptor.iProduct; break;
+	case LIBUSB_DEVICE_STRING_SERIAL_NUMBER: string_descriptor_idx = dev->device_descriptor.iSerialNumber; break;
+	default: return LIBUSB_ERROR_NOT_SUPPORTED;
```

Comment:
Shouldn't this be `LIBUSB_ERROR_INVALID_PARAM`?

Reply:
Sure.  Will change.